core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java

```diff
@@ -0,0 +1,431 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.metrics.stats.multifield;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Descriptive stats gathered per shard. Coordinating node computes final pearson product coefficient
+ * based on these descriptive stats
+ *
+ * @internal
+ */
+class MultiFieldStatsResults implements Streamable {
+    /** count of observations (same number of observations per field) */
+    protected long docCount;
+    /** per field sum of observations */
+    protected HashMap<String, Double> fieldSum;
+    /** counts */
+    protected HashMap<String, Long> counts;
+    /** mean values */
+    protected HashMap<String, Double> means;
+    /** variance values */
+    protected HashMap<String, Double> variances;
+    /** skewness values */
+    protected HashMap<String, Double> skewness;
+    /** kurtosis values */
+    protected HashMap<String, Double> kurtosis;
+    /** covariance values */
+    protected HashMap<String, HashMap<String, Double>> covariances;
+    /** pearson product correlation coefficients */
+    protected HashMap<String, HashMap<String, Double>> correlation;
+
+    /** cache flag to avoid recompute if nothing has been added */
+    private boolean dirty;
+
+    /**
+     * Base ctor
+     */
+    public MultiFieldStatsResults() {
+        docCount = 0;
+        dirty = true;
+        counts = new HashMap<>();
+        fieldSum = new HashMap<>();
+        means = new HashMap<>();
+        skewness = new HashMap<>();
+        kurtosis = new HashMap<>();
+        covariances = new HashMap<>();
+        variances = new HashMap<>();
+    }
+
+    /**
+     * Returns covariance between two fields
+     */
+    public double getCovariance(String fieldX, String fieldY) {
+        if (fieldX.equals(fieldY)) {
+            return variances.get(fieldX);
+        } else if (variances.containsKey(fieldX) == false && variances.containsKey(fieldY) == false) {
```

Comment:
Should this be an OR?  E.g. if either of the fields are missing, rather than both?  Looks like it could cause problems if one field is missing when it gets to `getValFromUpperTriangularMatrix()`?